### PR TITLE
Implement plot_fst_gwss function

### PR DESCRIPTION
**Pull request by sanjaynagi**
Adding the public functions `fst_gwss()` and `plot_fst_gwss()`, which calculate and plot hudson's Fst across a contig. 

The parameters (and code generally) mirror that of the `h1x_gwss()` functions, apart from loading genotypes over haplotypes and the `analysis` parameter is replaced with `site_mask`. 

Should there be a default for `site_mask` here? If we did have a default, I guess it would be `None` or `'gamb_colu_arab'`? Currently the default is `None`.

Ive also added a notebook and a test (the test mirrors the H1X test).

### CODE REVIEW

**Comment by sanjaynagi** [comment]
mentioned in #293

**Comment by sanjaynagi** [comment]
Thinking about it, I don't think there should be a default for `site_mask`. Have updated accordingly.

**Comment by sanjaynagi** [comment]
heres a colab[ notebook](https://colab.research.google.com/drive/1veneHIpBWV3vu-fFczecsXJ3WEHg6n4B) :)

File: malariagen_data/ag3.py [hunk 1]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            random_seed=random_seed,
+        )
+
+        print("computing allele counts for cohort 1")
+        gt1 = allel.GenotypeDaskArray(ds_snps1["call_genotype"].data)
+        ac1 = gt1.count_alleles()
+
+        print("computing allele counts for cohort 2")
+        gt2 = allel.GenotypeDaskArray(ds_snps2["call_genotype"].data)
+        ac2 = gt2.count_alleles()
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6588]:
```suggestion
        gt1 = allel.GenotypeDaskArray(ds_snps1["call_genotype"].data)
        with self._dask_progress(desc="Compute allele counts for cohort 1"):
            ac1 = gt1.count_alleles().compute()

        gt2 = allel.GenotypeDaskArray(ds_snps2["call_genotype"].data)
        with self._dask_progress(desc="Compute allele counts for cohort 2"):
            ac2 = gt2.count_alleles().compute()
```

Use `_dask_progress()` here to get a tqdm progress bar. Also make sure to call `.compute()` to actually trigger the computation at this point.

File: malariagen_data/ag3.py [hunk 2]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+
+        return results
+
+    def fst_gwss(
+        self,
+        contig,
+        window_size,
+        cohort1_query,
+        cohort2_query,
+        site_mask,
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6605]:
```suggestion
        site_mask="gamb_colu_arab",
```

Suggest defaulting to safest site mask.

File: malariagen_data/ag3.py [hunk 3]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        cohort2_query : str
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        sample_sets : str or list of str, optional
+            Can be a sample set identifier (e.g., "AG1000G-AO") or a list of
+            sample set identifiers (e.g., ["AG1000G-BF-A", "AG1000G-BF-B"]) or a
+            release identifier (e.g., "3.0") or a list of release identifiers.
+        site_mask : {"gamb_colu_arab", "gamb_colu", "arab"}
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6628]:
```suggestion
        site_mask : {"gamb_colu_arab", "gamb_colu", "arab"}, optional
```

File: malariagen_data/ag3.py [hunk 4]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            Site filters mask to apply.
+        cohort_size : int, optional
+            If provided, randomly down-sample to the given cohort size.
+        random_seed : int, optional
+            Random seed used for down-sampling.
+        Returns
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6634]:
```suggestion

        Returns
```

Add newline.

File: malariagen_data/ag3.py [hunk 5]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+        random_seed=42,
+        title=None,
+        width=DEFAULT_GENOME_PLOT_WIDTH,
+        height=200,
+        show=True,
+        x_range=None,
+    ):
+        """Run and plot a Fst genome-wide scan to detect genome regions
+        with shared selective sweeps between two cohorts.
+        Parameters
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6686]:
```suggestion
        """Run and plot a Fst genome-wide scan to investigate genetic 
        differentiation between two cohorts.

        Parameters
```

File: malariagen_data/ag3.py [hunk 6]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+
+        return x, fst
+
+    def plot_fst_gwss_track(
+        self,
+        contig,
+        window_size,
+        cohort1_query,
+        cohort2_query,
+        site_mask,
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6674]:
```suggestion
        site_mask="gamb_colu_arab",
```

File: malariagen_data/ag3.py [hunk 7]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        cohort2_query : str
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        sample_sets : str or list of str, optional
+            Can be a sample set identifier (e.g., "AG1000G-AO") or a list of
+            sample set identifiers (e.g., ["AG1000G-BF-A", "AG1000G-BF-B"]) or a
+            release identifier (e.g., "3.0") or a list of release identifiers.
+        site_mask : {"gamb_colu_arab", "gamb_colu", "arab"}
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6702]:
```suggestion
        site_mask : {"gamb_colu_arab", "gamb_colu", "arab"}, optional
```

File: malariagen_data/ag3.py [hunk 8]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+        cohort_size=30,
+        random_seed=42,
+        title=None,
+        width=DEFAULT_GENOME_PLOT_WIDTH,
+        track_height=190,
+        genes_height=DEFAULT_GENES_TRACK_HEIGHT,
+    ):
+        """Run and plot an Fst genome-wide scan to detect genome regions
+        with shared selective sweeps between two cohorts.
+        Parameters
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6798]:
```suggestion
        """Run and plot an Fst genome-wide scan to investigate genetic
        differentiation between two cohorts.
        
        Parameters
```

File: malariagen_data/ag3.py [hunk 9]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            Random seed used for down-sampling.
+        title : str, optional
+            If provided, title string is used to label plot.
+        width : int, optional
+            Plot width in pixels (px).
+        track_height : int. optional
+            GWSS track height in pixels (px).
+        genes_height : int. optional
+            Gene track height in pixels (px).
+        Returns
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6826]:
```suggestion

        Returns
```

File: malariagen_data/ag3.py [hunk 10]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+
+        return fig
+
+    def plot_fst_gwss(
+        self,
+        contig,
+        window_size,
+        cohort1_query,
+        cohort2_query,
+        site_mask,
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6787]:
```suggestion
        site_mask="gamb_colu_arab",
```

File: malariagen_data/ag3.py [hunk 11]
```diff
@@ -6549,6 +6549,323 @@ def displayTapNodeData(data):
...
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        cohort2_query : str
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        sample_sets : str or list of str, optional
+            Can be a sample set identifier (e.g., "AG1000G-AO") or a list of
+            sample set identifiers (e.g., ["AG1000G-BF-A", "AG1000G-BF-B"]) or a
+            release identifier (e.g., "3.0") or a list of release identifiers.
+        cohort_size : int, optional
```
Comment (alimanfoo) [inline malariagen_data/ag3.py:6814]: Missing site_mask parameter in docstring.

File: tests/test_ag3.py [hunk 12]
```diff
@@ -3076,3 +3076,37 @@ def test_haplotype_joint_frequencies():
...
+    assert x.ndim == fst.ndim == 1
+    assert x.shape == fst.shape
+
+    # check some values
+    assert_allclose(x[0], 56835.9649, rtol=1e-5), x[0]
+    assert_allclose(fst[0], 0.0405522778148594, rtol=1e-5), fst[0]
+    assert np.all(fst <= 1)
+    assert np.all(
+        fst >= 0
+    )  # not entirely sure about this one because Fst can sometimes be below 0. But in the example used, it is never below 0.
```
Comment (alimanfoo) [inline tests/test_ag3.py:3112]: Suggest to check values are within expected range, e.g., -0.1 to 1.0.

**Review by alimanfoo — COMMENTED** [review]
Thanks @sanjaynagi, looks great, just a few nits.

**Comment by alimanfoo** [comment]
> Thinking about it, I don't think there should be a default for `site_mask`. Have updated accordingly.

I think we do default to the "gamb_colu_arab" site mask elsewhere, e.g., when running a PCA. Probably reasonable to do so here I reckon.

**Comment by alimanfoo** [comment]
ping @sanjaynagi

**Comment by sanjaynagi** [comment]
all should be resolved, but the dask progress bar is appearing twice for each cohort, any idea why?

image

**Comment by alimanfoo** [comment]
Thanks so much @sanjaynagi! It looks like the double progress bar is down to the fact that if you call `GenotypeDaskArray.count_alleles()` without specifying a `max_allele` parameter, then there is an additional computation to first scan the array to find the maximum allele value. The context manager picks up and reports both that maximum value computation and the subsequent allele count computation. I took the liberty to push a fix to your branch. OK for this to be merged?

**Comment by sanjaynagi** [comment]
Awesome, thanks for doing that. Yeah, go ahead!

**Review by alimanfoo — APPROVED** [review]

**Pull request merged**